bookkeeper-server/src/main/java/org/apache/bookkeeper/client/RackawareEnsemblePlacementPolicyImpl.java

```diff
@@ -414,21 +414,52 @@ protected PlacementResult<List<BookieId>> newEnsembleInternal(
                 }
                 return PlacementResult.of(addrs, PlacementPolicyAdherence.FAIL);
             }
-
-            for (int i = 0; i < ensembleSize; i++) {
-                String curRack;
-                if (null == prevNode) {
-                    if ((null == localNode) || defaultRack.equals(localNode.getNetworkLocation())) {
-                        curRack = NodeBase.ROOT;
+            try {
+                //Choose different rack nodes.
+                String curRack = null;
+                for (int i = 0; i < ensembleSize; i++) {
+                    if (null == prevNode) {
+                        if ((null == localNode) || defaultRack.equals(localNode.getNetworkLocation())) {
+                            curRack = NodeBase.ROOT;
+                        } else {
+                            curRack = localNode.getNetworkLocation();
+                        }
                     } else {
-                        curRack = localNode.getNetworkLocation();
+                        if (!curRack.startsWith("~")) {
+                            curRack = "~" + prevNode.getNetworkLocation();
+                        } else {
+                            curRack = curRack + NetworkTopologyImpl.NODE_SEPARATOR + prevNode.getNetworkLocation();
+                        }
                     }
-                } else {
-                    curRack = "~" + prevNode.getNetworkLocation();
+                    prevNode = selectRandomFromRack(curRack, excludeNodes, ensemble, ensemble);
```

Comment:
Can we set `fallbackRandom` flag in `selectRandomFromRack` to support random selection instead of random select in BKNotEnoughBookiesException